src/util/taproot.rs

```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    fn insert(mut self, mut node: NodeInfo, mut depth: usize) -> Result<Self, TaprootBuilderError> {
+        // early error on invalid depth. Though this will be checked later
+        // while constructing TaprootMerkelBranch
+        if depth > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            return Err(TaprootBuilderError::InvalidMerkleTreeDepth(depth));
+        }
+        // We cannot insert a leaf at a lower depth while a deeper branch is unfinished. Doing
+        // so would mean the add_leaf/add_hidden invocations do not correspond to a DFS traversal of a
+        // binary tree.
+        if depth + 1 < self.branch.len() {
+            return Err(TaprootBuilderError::NodeNotInDfsOrder);
+        }
+
+        while self.branch.len() == depth + 1 {
+            let child = match self.branch.pop() {
+                None => unreachable!("Len of branch checked to be >= 1"),
+                Some(Some(child)) => child,
+                // Needs an explicit push to add the None that we just popped.
+                // Cannot use .last() because of borrow checker issues.
+                Some(None) => {
+                    self.branch.push(None);
+                    break;
+                } // Cannot combine further
+            };
+            if depth == 0 {
+                // We are trying to combine two nodes at root level.
+                // Can't propagate further up than the root
+                return Err(TaprootBuilderError::OverCompleteTree);
+            }
+            node = NodeInfo::combine(node, child)?;
+            // Propagate to combine nodes at a lower depth
+            depth -= 1;
+        }
+
+        if self.branch.len() < depth + 1 {
+            // add enough nodes so that we can insert node at depth `depth`
+            let num_extra_nodes = depth + 1 - self.branch.len();
+            self.branch
+                .extend((0..num_extra_nodes).into_iter().map(|_| None));
+        }
+        // Push the last node to the branch
+        self.branch[depth] = Some(node);
+        Ok(self)
+    }
+}
+
+// Internally used structure to represent the node information in taproot tree
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct NodeInfo {
+    /// Merkle Hash for this node
+    hash: sha256::Hash,
+    /// information about leaves inside this node
+    leaves: Vec<LeafInfo>,
+}
+
+impl NodeInfo {
+    // Create a new NodeInfo with omitted/hidden info
+    fn new_hidden(hash: sha256::Hash) -> Self {
+        Self {
+            hash: hash,
+            leaves: vec![],
+        }
+    }
+
+    // Create a new leaf with NodeInfo
+    fn new_leaf_with_ver(script: Script, ver: LeafVersion) -> Self {
+        let leaf = LeafInfo::new(script, ver);
+        Self {
+            hash: leaf.hash(),
+            leaves: vec![leaf],
+        }
+    }
+
+    // Combine two NodeInfo's to create a new parent
+    fn combine(a: Self, b: Self) -> Result<Self, TaprootBuilderError> {
```

Comment:
Why the function does not take `&mut self` as first argument? This will make calling it above in line https://github.com/rust-bitcoin/rust-bitcoin/pull/677/files#diff-aa9330b8acaba59f1d0418de0acdd5836506b48566d28886e2a3558e1e19de1bR467 simpler: `node.combine(child)?;`

Reply:
I guess this is about how we view the tree structure. I think conceptually it makes more sense to consume both the children and create a parent. I believe this helps in understanding the function source code easily whereas the other might be slightly more complicated to reason about.